### PR TITLE
Fix SystemLog

### PR DESCRIPTION
**Pull request by erikzhang**

### CODE REVIEW

**Comment by shargon** [comment]
I think that both solutions are fine, this is the protection and https://github.com/neo-project/neo/pull/1571 it's the fix

File: src/SystemLog/Logger.cs [hunk 1]
```diff
@@ -39,6 +40,9 @@ void ILogPlugin.Log(string source, LogLevel level, string message)
 
                 if (!string.IsNullOrEmpty(Settings.Default.Path))
                 {
+                    StringBuilder sb = new StringBuilder(source);
+                    foreach (char c in GetInvalidFileNameChars())
+                        sb.Replace(c, '-');
```
Comment (Tommo-L) [inline src/SystemLog/Logger.cs:45]: Throw exception or fix in neo-core? I think we shouldn't modify the `source`, which is the log file name. 🤔

File: src/SystemLog/Logger.cs [hunk 2]
```diff
@@ -39,6 +40,9 @@ void ILogPlugin.Log(string source, LogLevel level, string message)
...
                     var path = Combine(Settings.Default.Path, source);
```
Comment (shargon) [inline src/SystemLog/Logger.cs:46]:
```suggestion
                    var path = Combine(Settings.Default.Path, sb.ToString());
```

File: src/SystemLog/Logger.cs (hunk 1 above)
Comment (shargon) [reply src/SystemLog/Logger.cs]: I think that we should merge both

Comment (erikzhang) [reply src/SystemLog/Logger.cs]: I think we should only merge this one. `source` is not filename. We use it as filename in SystemLog, so we fix it in SystemLog.

Comment (Tommo-L) [reply src/SystemLog/Logger.cs]: The problem is, user want a log file named `A`, but we create `B`, this is unreasonable.

**Comment by erikzhang** [comment]
@shargon `StringBuilder` is more efficient.

File: src/SystemLog/Logger.cs (hunk 1 above)
Comment (erikzhang) [reply src/SystemLog/Logger.cs:45]: As I have said, the parameter `source` is not a filename. It is the log source. We just use it as filename in SystemLog.

Comment (Tommo-L) [reply src/SystemLog/Logger.cs:45]: Got.

**Comment by shargon** [comment]
But `Replace` returns a new `StringBuilder`, that's why I thought that in this case maybe it's not better.

File: src/SystemLog/Logger.cs [hunk 3]
```diff
@@ -39,7 +40,10 @@ void ILogPlugin.Log(string source, LogLevel level, string message)
 
                 if (!string.IsNullOrEmpty(Settings.Default.Path))
                 {
-                    var path = Combine(Settings.Default.Path, source);
+                    StringBuilder sb = new StringBuilder(source);
+                    foreach (char c in GetInvalidFileNameChars())
+                        sb.Replace(c, '-');
```
Comment (shargon) [inline src/SystemLog/Logger.cs:45]:
```suggestion
                        sb = sb.Replace(c, '-');
```

Comment (erikzhang) [reply src/SystemLog/Logger.cs:45]: It doesn't returns new `StringBuilder`. It returns the old one. There will be no new one be created.

**Pull request merged**